#' infer the time calibrated phylogeny associated with the
#' @param alignment Phydat object containing the focal alignment
#' @param treatment_name string to be appended to BEAST files
#' @param inference_model inference model used in BEAST analysis,
#' as generated by the function beautier::create_inference_model
#' @param mcmc_seed seed of the mcmc chain, default is the system time
#' @param burnin burnin of posterior distribution
#' @param working_dir beast2 working dir
#' @return list with all trees, and the consensus tree
#' @export
infer_phylogeny <- function(alignment,
                            treatment_name,
                        inference_model = beautier::create_inference_model(
                          site_model = beautier::create_jc69_site_model(),
                          clock_model = beautier::create_strict_clock_model(),
                          tree_prior = beautier::create_yule_tree_prior(),
                          mcmc = beautier::create_mcmc(chain_length = 1e6,
                                                       store_every = 5000)
                        ),
                            mcmc_seed = NULL,
                            burnin,
                            working_dir = NULL)  {

  if (is.null(working_dir)) working_dir <- getwd()

  temp_file_name <- "temp.fasta"
  phangorn::write.phyDat(alignment, file = temp_file_name, format = "fasta")

  if(is.null(mcmc_seed)) mcmc_seed = round(as.numeric(Sys.time()))


  intended_options <-  beastier::create_beast2_options(
    overwrite = TRUE,
    beast2_working_dir = getwd(),
    output_trees_filenames = paste0(treatment_name, ".trees"),
    output_log_filename = paste0(treatment_name, ".log"),
    rng_seed = mcmc_seed,
    verbose = TRUE
  )

  cat(intended_options$beast2_working_dir, "\n")
  intended_options <- peregrine::to_pff_beast2_options(intended_options)
  cat("working dir updated to: \n")
  cat(intended_options$beast2_working_dir, "\n")


  posterior <- babette::bbt_run_from_model(
    temp_file_name,
    inference_model,
    beast2_options = intended_options
  )

  file.remove(temp_file_name)

  beast_log <- tracerer::remove_burn_ins(
    posterior$estimates,
    burn_in_fraction = burnin
  )

  esses <- tracerer::calc_esses(beast_log, sample_interval = 5000)
  if (min(esses) < 200) {
    warning("WARNING! MCMC chain not converged!\n")
  }

  found_trees <- tracerer::parse_beast_trees(paste0(treatment_name, ".trees"))
  remaining <- floor(burnin * length(found_trees))
  found_trees <- found_trees[remaining:length(found_trees)]

  consensus_tree <-  phangorn::maxCladeCred(found_trees)

  output <- list("all_trees" = found_trees,
                 "mcc_tree" = consensus_tree)

  return(output)
}
